go/private/binary.bzl

```diff
@@ -33,11 +35,32 @@ def _go_binary_impl(ctx):
     gc_linkopts=gc_linkopts(ctx),
     x_defs=ctx.attr.x_defs)
 
-  return struct(
-      files = depset([ctx.outputs.executable]),
-      runfiles = lib_result.runfiles,
-      cgo_object = lib_result.cgo_object,
-  )
+  # Static linking (in the 'static' output group)
+  static_linkopts = [
+      "-linkmode", "external",
+      "-extldflags", "-static",
+  ]
+  static_executable = ctx.new_file(ctx.attr.name + ".static")
+  emit_go_link_action(
+    ctx,
+    transitive_go_libraries=lib_result.transitive_go_libraries,
+    transitive_go_library_paths=lib_result.transitive_go_library_paths,
+    cgo_deps=lib_result.transitive_cgo_deps,
+    libs=lib_result.files,
+    executable=static_executable,
+    gc_linkopts=gc_linkopts(ctx) + static_linkopts,
+    x_defs=ctx.attr.x_defs)
+
+  return [
+      DefaultInfo(
+          files = depset([ctx.outputs.executable]),
+          runfiles = lib_result.runfiles,
+          cgo_object = lib_result.cgo_object,
+      ),
+      OutputGroupInfo(
+          static = depset([static_executable]),
+      ),
```

Comment:
Minor points for discussion, could always be done in a different change later...

Should we also have a named group for the dynamic form so you can be explicit about which one you want?
What should we do on platforms where static is not possible? 
* Not declare the static group
* have the static depset empty
* leave it failing when trying to build

Should we have a way of allowing rules to use the static if available, and the dynamic otherwise, or are the consuming rules always going to have to modify their behaviour anyway?